pytorch_lightning/trainer/data_loading.py

```diff
@@ -124,23 +124,25 @@ def _worker_check(self, dataloader: DataLoader, name: str) -> None:
         # ddp_spawn + num_workers > 0 don't mix! tell the user
         is_dataloader = isinstance(dataloader, DataLoader)
         using_spawn = self.distributed_backend == 'ddp_spawn'
-        if is_dataloader and dataloader.num_workers > 0 and not on_windows and using_spawn:
+        if on_windows:
+            pass
+        elif is_dataloader and dataloader.num_workers > 0 and using_spawn:
             rank_zero_warn('Dataloader(num_workers>0) and ddp_spawn do not mix well! '
                            'Your performance might suffer dramatically. '
                            'Please consider setting distributed_backend=ddp to use num_workers > 0 '
                            '(this is a bottleneck of Python .spawn() and PyTorch')
 
-        elif is_dataloader and dataloader.num_workers <= 2 and not on_windows and not using_spawn:
+        elif is_dataloader and dataloader.num_workers == 0 and using_spawn:
+            rank_zero_warn('You are using `distributed_backend=ddp_spawn` with num_workers=0. '
+                           'For much faster performance, switch to `distributed_backend=ddp` and set `num_workers>0`')
+
+        elif is_dataloader and dataloader.num_workers <= 2 and multiprocessing.cpu_count() > 2 and not using_spawn:
```

Comment:
Why is `2` taken as the minimum threshold for this warning? Why not `0`?

Reply:
The goal is to check if too few workers are assigned for the dataloader. If you only have 2 workers on a machine with 8 cores, you'll still be underutilized. We could consider only warning if the machine has >=4 cores. 

cc @PyTorchLightning/core-contributors